src/gazebo_gps_plugin.cpp

```diff
@@ -0,0 +1,200 @@
+
+// The home position can be specified using the environment variables:
+// PX4_HOME_LAT, PX4_HOME_LON, and PX4_HOME_ALT
+
+// Zurich Irchel Park
+static double lat_home = 47.397742 * M_PI / 180;  // rad
+static double lon_home = 8.545594 * M_PI / 180;  // rad
+static double alt_home = 488.0; // meters
+// Seattle downtown (15 deg declination): 47.592182, -122.316031
+// static const double lat_home = 47.592182 * M_PI / 180;  // rad
+// static const double lon_home = -122.316031 * M_PI / 180;  // rad
+// static const double alt_home = 86.0; // meters
+static const double earth_radius = 6353000;  // m
+
+GpsPlugin::GpsPlugin() : ModelPlugin()
+{
+    
+}
+
+GpsPlugin::~GpsPlugin()
+{
+  event::Events::DisconnectWorldUpdateBegin(updateConnection_);
+}
+
+void GpsPlugin::Load(physics::ModelPtr _model, sdf::ElementPtr _sdf)
+{
+  // Store the pointer to the model.
+  model_ = _model;
+
+  world_ = model_->GetWorld();
+
+  // Use environment variables if set for home position.
+  const char *env_lat = std::getenv("PX4_HOME_LAT");
+  const char *env_lon = std::getenv("PX4_HOME_LON");
+  const char *env_alt = std::getenv("PX4_HOME_ALT");
+  
+  // Get noise param
+  if(_sdf->HasElement("gpsNoise"))
+  {
+    getSdfParam<bool>(_sdf, "gpsNoise", gps_noise_, gps_noise_);
+  }
+  
+  else {
+    gps_noise_ = true;    
+  }
+
+  if (env_lat) {
+    gzmsg << "Home latitude is set to " << env_lat << ".\n";
+    lat_home = std::stod(env_lat) * M_PI / 180.0;
+  }
+  if (env_lon) {
+    gzmsg << "Home longitude is set to " << env_lon << ".\n";
+    lon_home = std::stod(env_lon) * M_PI / 180.0;
+  }
+  if (env_alt) {
+    gzmsg << "Home altitude is set to " << env_alt << ".\n";
+    alt_home = std::stod(env_alt);
+  }
+  
+  namespace_.clear();
+  if (_sdf->HasElement("robotNamespace")) {
+    namespace_ = _sdf->GetElement("robotNamespace")->Get<std::string>();
+  } else {
+    gzerr << "[gazebo_mavlink_interface] Please specify a robotNamespace.\n";
+  }
+
+  node_handle_ = transport::NodePtr(new transport::Node());
+  node_handle_->Init(namespace_);
+  
+   // Listen to the update event. This event is broadcast every
+  // simulation iteration.
+  updateConnection_ = event::Events::ConnectWorldUpdateBegin(
+      boost::bind(&GpsPlugin::OnUpdate, this, _1));
+  
+  last_gps_time_ = world_->GetSimTime();
+  gps_update_interval_ = 0.2;  // in seconds for 5Hz
+  gps_delay_ = 0.12; // in seconds
+  gravity_W_ = world_->GetPhysicsEngine()->GetGravity();
+    
+  gps_pub_ = node_handle_->Advertise<gps_msgs::msgs::Gps>("~/" + model_->GetName() + "/gps", 10);
+}
+
+void GpsPlugin::OnUpdate(const common::UpdateInfo&){
+    
+  common::Time current_time = world_->GetSimTime();
+  double dt = (current_time - last_gps_time_).Double();
+    
+  math::Pose T_W_I = model_->GetWorldPose(); //TODO(burrimi): Check tf.
+  math::Vector3& pos_W_I = T_W_I.pos;  // Use the models' world position for GPS and pressure alt.
+  
+  reproject(pos_W_I, lat_rad, lon_rad);
```

Comment:
so in my understanding these are the `lat_rad` and `lon_rad` from `gazebo_gps_plugin.h`, which means that they actually are just used in `gazebo_gps_plugin` to compute the data fields of the `Gps.proto`. What about the `lat_rad` and `lon_rad` of the IMU callback? They will stay as zero cause it's the value they are init on the constructor and they are never updated. So you should update the `lat_rad` and `lon_rad` of `gazebo_mavlink_interface` on the GPS callback:
```
lat_rad = gps_msg->lat_rad();
lon_rad = gps_msg->lon_rad();
```